packages/fluentui/react-northstar/src/utils/positioner/usePopper.ts

```diff
@@ -0,0 +1,391 @@
+  const arrowRef = useCallbackRef<HTMLElement | null>(null, handlePopperUpdate, true);
+
+  React.useImperativeHandle(
+    options.popperRef,
+    () => ({
+      updatePosition: () => {
+        popperInstanceRef.current?.update();
+      },
+    }),
+    [],
+  );
+
+  useIsomorphicLayoutEffect(() => {
+    handlePopperUpdate();
+
+    return () => {
+      popperInstanceRef.current?.destroy();
+      popperInstanceRef.current = null;
+    };
+  }, [options.enabled]);
+  useIsomorphicLayoutEffect(() => {
+    if (!isFirstMount) {
+      popperInstanceRef.current?.setOptions(
+        resolvePopperOptions(targetRef.current, containerRef.current, arrowRef.current),
+      );
+    }
+  }, [resolvePopperOptions]);
+
+  if (process.env.NODE_ENV !== 'production') {
+    // This checked should run only in development mode
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    React.useEffect(() => {
+      if (containerRef.current) {
+        const contentNode = containerRef.current;
+        const treeWalker = contentNode.ownerDocument?.createTreeWalker(contentNode, NodeFilter.SHOW_ELEMENT, {
+          acceptNode: hasAutofocusFilter,
+        });
+
+        while (treeWalker.nextNode()) {
+          const node = treeWalker.currentNode;
+          // eslint-disable-next-line no-console
+          console.warn('<Popper>:', node);
+          // eslint-disable-next-line no-console
+          console.warn(
+            [
+              '<Popper>: ^ this node contains "autoFocus" prop on a React element. This can break the initial',
+              'positioning of an element and cause a window jump effect. This issue occurs because React polyfills',
+              '"autoFocus" behavior to solve inconsistencies between different browsers:',
+              'https://github.com/facebook/react/issues/11851#issuecomment-351787078',
+              '\n',
+              'However, ".focus()" in this case occurs before any other React effects will be executed',
+              '(React.useEffect(), componentDidMount(), etc.) and we can not prevent this behavior. If you really',
+              'want to use "autoFocus" please add "position: fixed" to styles of the element that is wrapped by',
+              '"Popper".',
+              `In general, it's not recommended to use "autoFocus" as it may break accessibility aspects:`,
+              'https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-autofocus.md',
+              '\n',
+              'We suggest to use the "trapFocus" prop on Fluent components or a catch "ref" and then use',
+              '"ref.current.focus" in React.useEffect():',
+              'https://reactjs.org/docs/refs-and-the-dom.html#adding-a-ref-to-a-dom-element',
+            ].join(' '),
+          );
+        }
+      }
+      // We run this check once, no need to add deps here
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, []);
+  }
+
+  return [targetRef, containerRef, arrowRef];
```

Comment:
This seems weird, especially since you return an array of three of the same type. Comments on array type returns don't give any kind of TS hint

![image](https://user-images.githubusercontent.com/20744592/110255903-091e5b00-7f8e-11eb-869c-78ad46955fc9.png)

Reply:
https://github.com/microsoft/fluentui/blob/3b422dadd4809698b819426f5b4f2de31253bf02/packages/fluentui/react-northstar/src/utils/positioner/usePopper.ts#L263-L268

Initially, I had better typings, but then `refs` are not assignable to elements:

```tsx
  const ref = usePopper()
  // 🔴 Type 'MutableRefObject<HTMLElement>' is not assignable to type 'LegacyRef<HTMLDivElement> | undefined'.
  return <div ref={ref} />
```

[TypeScript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQLABQ9BArgHb7AQtxOpYAKEMGCxQAFAEoAXJiowAdAFkmMZACMANlmwEA8qoBWWfAB4AEgBUFAGQCimkFhYwAfHADe9OHEowmULth48jxaWARmlrb2ji6iACYQuEwOTnK4lMgwWHZYKTCiZHHAAG5k4uL0AL70jKzsnHAAwiSQLDES7p5wuJyo8JQEcAC83LwCQiISNXRePn5cxkXF3mFDbgOVcAD0zlXTPSyoEJpy6hAA5qLN4Jzt5UA)

---

I see following options:
- fix typings (how?)
- keep as it is
- return an object instead of an array, it will improve naming, but typings will still be awful

Opinions? Other ideas? 😃

Reply:
What about `<unknown>` and force the user to cast it 
[modifed ts playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQLABQ9BArgHb7AQtxOpYAKEMGCxQAFAEoAXJiowAdAFkmMZACMANlmwEA8qoBWWfAB5WAaxYQA7iwB8cAN704cSjCZQu2PPJ5asBMYAEgAqCgAyAKKaIFgsMLaiACYQuEyx8XK4lMgwWNFYGTCiZEnAAG5k4uL0AL70jKzsnHAAwiSQLHHF4o7OcLicqPCUBHAAvNy8AkIiEijo3viKymqa2nqGJqERACIVBUW2DXQubh5cxmXlrgHjDqO1cAD0x3T1DHSDLKgQmnLqCAAc1E7XAnG6EnEQA)

Reply:
> unknown

That will force users to as case every usage which seems a bit overkill

> fix types
Best solution I could come up within a few minutes would be to use generics so that the user can pass the correct HTML element types similar to what `useRef` will do.

> return an object instead of array
IMO that should be a must... consumers can still change naming during destructuring, and in a case where an array returns three of the same type object it might not be so clear which ref is being used

Reply:
@ling1726 now it returns an object (5cbb1ed):

```tsx
type T = {
  targetRef: React.MutableRefObject<any>;
  containerRef: React.MutableRefObject<any>;
  arrowRef: React.MutableRefObject<any>;
};
```

I would like to have better typings there anyway. I tried to use generics, but it will be still `any` otherwise TS screams...

Reply:
I think that it is an improvement 😄

Reply:
I would say it's definitely worth adding tsdoc to those return properties... since IDEs will show that description in hints